File: apps/web/app/services/api/background.ts
// Background Task API Service

import { ApiService } from './base';

export interface RefreshRequest {
  mode: 'smart' | 'full' | 'minimal';
}

export interface ComputeRequest {
  momentum_weight?: number;
  market_cap_weight?: number;
  risk_parity_weight?: number;
}

export interface ReportRequest {
  report_type: 'performance' | 'allocation' | 'risk';
  period_days: number;
}

export interface CleanupRequest {
  days_to_keep: number;
}

export interface TaskResponse {
  task_id: string;
  status: string;
  message: string;
}

export interface TaskStatus {
  task_id: string;
  status: 'PENDING' | 'STARTED' | 'SUCCESS' | 'FAILURE' | 'RETRY' | 'REVOKED';
  result?: any;
  error?: string;
  traceback?: string;
  current?: number;
  total?: number;
}

export interface ActiveTasks {
  active: Record<string, any[]>;
  scheduled: Record<string, any[]>;
  reserved: Record<string, any[]>;
  stats: {
    total_active: number;
    total_scheduled: number;
    total_reserved: number;
  };
}

export class BackgroundTaskService extends ApiService {
  async triggerRefresh(request: RefreshRequest): Promise<TaskResponse> {
    return this.post<TaskResponse>('/api/v1/background/refresh', request);
  }

  async triggerCompute(request: ComputeRequest): Promise<TaskResponse> {
    return this.post<TaskResponse>('/api/v1/background/compute', request);
  }

  async generateReport(request: ReportRequest): Promise<TaskResponse> {
    return this.post<TaskResponse>('/api/v1/background/report', request);
  }

  async cleanupData(request: CleanupRequest): Promise<TaskResponse> {
    return this.post<TaskResponse>('/api/v1/background/cleanup', request);
  }

  async getTaskStatus(taskId: string): Promise<TaskStatus> {
    return this.get<TaskStatus>(`/api/v1/background/status/${taskId}`);
  }

  async getActiveTasks(): Promise<ActiveTasks> {
    return this.get<ActiveTasks>('/api/v1/background/active');
  }

  // Polling helper for task completion
  async pollTaskStatus(
    taskId: string, 
    onProgress?: (status: TaskStatus) => void,
    intervalMs: number = 1000,
    maxAttempts: number = 300
  ): Promise<TaskStatus> {
    let attempts = 0;
    
    return new Promise((resolve, reject) => {
      const poll = async () => {
        try {
          const status = await this.getTaskStatus(taskId);
          
          if (onProgress) {
            onProgress(status);
          }
          
          if (status.status === 'SUCCESS' || status.status === 'FAILURE') {
            resolve(status);
            return;
          }
          
          attempts++;
          if (attempts >= maxAttempts) {
            reject(new Error('Task polling timeout'));
            return;
          }
          
          setTimeout(poll, intervalMs);
        } catch (error) {
          reject(error);
        }
      };
      
      poll();
    });
  }
}

export const backgroundTaskService = new BackgroundTaskService();

File: apps/api/app/routers/background.py
"""Background tasks API router."""
from fastapi import APIRouter, Depends, HTTPException
from typing import Dict, Any, Optional
from pydantic import BaseModel

from ..utils.token_dep import get_current_user
from ..models.user import User
from ..tasks.background_tasks import (
    refresh_market_data,
    compute_index,
    generate_report,
    cleanup_old_data,
    get_task_status
)

router = APIRouter()


class TaskResponse(BaseModel):
    """Response model for task creation."""
    task_id: str
    status: str
    message: str


class RefreshRequest(BaseModel):
    """Request model for refresh task."""
    mode: str = "smart"  # smart, full, minimal


class ComputeRequest(BaseModel):
    """Request model for compute task."""
    momentum_weight: Optional[float] = None
    market_cap_weight: Optional[float] = None
    risk_parity_weight: Optional[float] = None


class ReportRequest(BaseModel):
    """Request model for report generation."""
    report_type: str = "performance"  # performance, allocation, risk
    period_days: int = 30


class CleanupRequest(BaseModel):
    """Request model for cleanup task."""
    days_to_keep: int = 365


@router.post("/refresh", response_model=TaskResponse)
def trigger_refresh_task(
    request: RefreshRequest,
    user: User = Depends(get_current_user)
) -> TaskResponse:
    """Trigger market data refresh in the background."""
    try:
        # Start the task
        task = refresh_market_data.delay(mode=request.mode)
        
        return TaskResponse(
            task_id=task.id,
            status="started",
            message=f"Market data refresh started in {request.mode} mode"
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to start task: {str(e)}")


@router.post("/compute", response_model=TaskResponse)
def trigger_compute_task(
    request: ComputeRequest,
    user: User = Depends(get_current_user)
) -> TaskResponse:
    """Trigger index computation in the background."""
    try:
        # Prepare strategy config if provided
        strategy_config = None
        if any([request.momentum_weight, request.market_cap_weight, request.risk_parity_weight]):
            strategy_config = {}
            if request.momentum_weight is not None:
                strategy_config["momentum_weight"] = request.momentum_weight
            if request.market_cap_weight is not None:
                strategy_config["market_cap_weight"] = request.market_cap_weight
            if request.risk_parity_weight is not None:
                strategy_config["risk_parity_weight"] = request.risk_parity_weight
        
        # Start the task
        task = compute_index.delay(strategy_config=strategy_config)
        
        return TaskResponse(
            task_id=task.id,
            status="started",
            message="Index computation started"
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to start task: {str(e)}")


@router.post("/report", response_model=TaskResponse)
def trigger_report_task(
    request: ReportRequest,
    user: User = Depends(get_current_user)
) -> TaskResponse:
    """Generate a report in the background."""
    try:
        # Start the task
        task = generate_report.delay(
            report_type=request.report_type,
            period_days=request.period_days
        )
        
        return TaskResponse(
            task_id=task.id,
            status="started",
            message=f"Report generation started for {request.report_type}"
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to start task: {str(e)}")


@router.post("/cleanup", response_model=TaskResponse)
def trigger_cleanup_task(
    request: CleanupRequest,
    user: User = Depends(get_current_user)
) -> TaskResponse:
    """Clean up old data in the background."""
    try:
        # Start the task
        task = cleanup_old_data.delay(days_to_keep=request.days_to_keep)
        
        return TaskResponse(
            task_id=task.id,
            status="started",
            message=f"Data cleanup started, keeping {request.days_to_keep} days"
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to start task: {str(e)}")


@router.get("/status/{task_id}")
def get_task_result(
    task_id: str,
    user: User = Depends(get_current_user)
) -> Dict[str, Any]:
    """Get the status and result of a background task."""
    return get_task_status(task_id)


@router.get("/active")
def get_active_tasks(user: User = Depends(get_current_user)) -> Dict[str, Any]:
    """Get list of active tasks."""
    try:
        from ..core.celery_app import celery_app
        
        # Get active tasks
        inspect = celery_app.control.inspect()
        active = inspect.active()
        scheduled = inspect.scheduled()
        reserved = inspect.reserved()
        
        return {
            "active": active or {},
            "scheduled": scheduled or {},
            "reserved": reserved or {},
            "stats": {
                "total_active": sum(len(tasks) for tasks in (active or {}).values()),
                "total_scheduled": sum(len(tasks) for tasks in (scheduled or {}).values()),
                "total_reserved": sum(len(tasks) for tasks in (reserved or {}).values())
            }
        }
    except Exception as e:
        return {
            "error": str(e),
            "message": "Celery worker may not be running"
        }